framework/mesh/MeshGenerator/SplitFileMeshGenerator.cc

```diff
@@ -0,0 +1,399 @@
+#include "SplitFileMeshGenerator.h"
+
+#include "data_types/byte_array.h"
+#include "utils/chi_utils.h"
+
+#include "mesh/MeshHandler/chi_meshhandler.h"
+#include "mesh/VolumeMesher/chi_volumemesher.h"
+#include "mesh/MeshContinuum/chi_meshcontinuum.h"
+
+#include "chi_log.h"
+#include "utils/chi_timer.h"
+
+#include "ChiObjectFactory.h"
+
+#include <filesystem>
+
+#define scint static_cast<int>
+
+namespace chi_mesh
+{
+
+RegisterChiObject(chi_mesh, SplitFileMeshGenerator);
+
+// ##################################################################
+chi::InputParameters SplitFileMeshGenerator::GetInputParameters()
+{
+  chi::InputParameters params = MeshGenerator::GetInputParameters();
+
+  params.SetGeneralDescription(
+    "Generates the mesh only on location 0, thereafter partitions the mesh"
+    " but instead of broadcasting the mesh to other locations it creates binary"
+    " mesh files for each location.");
+  params.SetDocGroup("doc_MeshGenerators");
+
+  params.AddOptionalParameter(
+    "num_partitions",
+    0,
+    "The number of partitions to generate. If zero will "
+    "default to the number of MPI processes. Is "
+    "ignored if the number of MPI processes > 1.");
+
+  params.AddOptionalParameter(
+    "split_mesh_dir_path",
+    "SplitMesh",
+    "Path of the directory to be created for containing the split meshes.");
+
+  params.AddOptionalParameter("split_file_prefix",
+                              "split_mesh",
+                              "Prefix to use for all split mesh files");
+
+  params.AddOptionalParameter(
+    "read_only",
+    false,
+    "Controls whether the split mesh is recreated or just read.");
+
+  return params;
+}
+
+// ##################################################################
+SplitFileMeshGenerator::SplitFileMeshGenerator(
+  const chi::InputParameters& params)
+  : MeshGenerator(params),
+    num_parts_(params.GetParamValue<int>("num_partitions")),
+    split_mesh_dir_path_(
+      params.GetParamValue<std::string>("split_mesh_dir_path")),
+    split_file_prefix_(params.GetParamValue<std::string>("split_file_prefix")),
+    read_only_(params.GetParamValue<bool>("read_only"))
+{
+}
+
+// ##################################################################
+void SplitFileMeshGenerator::Execute()
```

Comment:
can you point me to the location where I expect that rank-0 will release most of the memory it used for create the mesh and partition it? I am looking for where rank-0 is releasing memory. Thanks!!

Reply:
Sure. Below is the first part of the `Execute` method:
```c++
void SplitFileMeshGenerator::Execute()
{
  const int num_mpi = Chi::mpi.process_count;
  const int num_parts = num_mpi == 1 ? num_parts_ : num_mpi;

  if (Chi::mpi.location_id == 0 and (not read_only_))
  {
    //======================================== Execute all input generators
    // Note these could be empty
    std::unique_ptr<UnpartitionedMesh> current_umesh = nullptr;
    for (auto mesh_generator_ptr : inputs_)
    {
      auto new_umesh =
        mesh_generator_ptr->GenerateUnpartitionedMesh(std::move(current_umesh));
      current_umesh = std::move(new_umesh);
    }

    //======================================== Generate final umesh
    current_umesh = GenerateUnpartitionedMesh(std::move(current_umesh));

    Chi::log.Log() << "Writing split-mesh with " << num_parts << " parts";
    const auto cell_pids = PartitionMesh(*current_umesh, num_parts);
    WriteSplitMesh(cell_pids, *current_umesh, num_parts);
    Chi::log.Log() << "Split-mesh with " << num_parts
                   << " parts successfully created";
  } // if home location
```

The unique ptr `std::unique_ptr<UnpartitionedMesh> current_umesh` only exists within the scope of the if statement ` if (Chi::mpi.location_id == 0 and (not read_only_))`. The moment we leave that scope the unique_ptr destroys the mesh it was pointing to. Thereafter all processors join up at the next Barrier after which each one reads their piece from the respective file. ;-)